File: mainActivity/src/main/java/com/sierrawireless/avphone/task/ApplicationClient.java
package com.sierrawireless.avphone.task;

import android.util.Log;

import com.sierrawireless.avphone.model.AvPhoneApplication;
import com.sierrawireless.avphone.model.AvPhoneObjectData;

import net.airvantage.model.AirVantageException;
import net.airvantage.model.Application;
import net.airvantage.model.ApplicationData;
import net.airvantage.model.AvSystem;
import net.airvantage.model.Protocol;
import net.airvantage.model.User;
import net.airvantage.utils.IAirVantageClient;
import net.airvantage.utils.Utils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ApplicationClient implements IApplicationClient {
    private static final String TAG = "ApplicationClient";

    private IAirVantageClient client;
    private User currentUser = null;

    ApplicationClient(IAirVantageClient client) {
        this.client = client;
    }

    @Override
    public Application ensureApplicationExists() throws IOException, AirVantageException {
        Application application = getApplication();

        if (application == null) {
            Log.d(TAG, "ensureApplicationExists: Create new application");
            application = createApplication();
            setApplicationCommunication(application.uid);
        }
        Log.d(TAG, "ensureApplicationExists: application is " + application.uid);
        return application;
    }

    @Override
    public void setApplicationData(String applicationUid, ArrayList<AvPhoneObjectData> customData)
            throws IOException, AirVantageException {
        List<ApplicationData> data = AvPhoneApplication.createApplicationData(customData);
        client.setApplicationData(applicationUid, data);
    }

    protected Application getApplication() throws IOException, AirVantageException {
        List<Application> applications = client.getApplications(AvPhoneApplication.appType(getCurrentUsername()));
        return Utils.first(applications);
    }

    private String getCurrentUsername() {
        if (currentUser == null) {
            try {
                currentUser = client.getCurrentUser();
            } catch (final IOException | AirVantageException e) {
                return "";
            }
        }
        return currentUser.email;
    }

    @Override
    public Application createApplication() throws IOException, AirVantageException {
        Application application = AvPhoneApplication.createApplication(getCurrentUsername());
        return client.createApplication(application);
    }

    @Override
    public void setApplicationCommunication(String applicationUid) throws IOException, AirVantageException {
        List<Protocol> protocols = AvPhoneApplication.createProtocols();
        client.setApplicationCommunication(applicationUid, protocols);
    }

    @Override
    public void addApplication(AvSystem system, Application application) throws IOException, AirVantageException {
        client.updateSystem(systemWithApplication(system, application));
    }

    private AvSystem systemWithApplication(AvSystem system, Application appToAdd) {

        AvSystem res = new AvSystem();
        res.uid = system.uid;
        res.type = system.type;
        res.applications = new ArrayList<>();

        boolean appAlreadyLinked = false;

        if (system.applications != null) {
            for (Application systemApp : system.applications) {
                Application resApp = new Application();
                resApp.uid = systemApp.uid;
                res.applications.add(resApp);
                if (resApp.uid.equals(appToAdd.uid)) {
                    appAlreadyLinked = true;
                }
            }
        }
        if (!appAlreadyLinked) {
            Application addedApp = new Application();
            addedApp.uid = appToAdd.uid;
            res.applications.add(addedApp);
        }

        return res;

    }

}

File: mainActivity/src/main/java/com/sierrawireless/avphone/task/ApplicationClient.kt
package com.sierrawireless.avphone.task

import android.util.Log
import com.sierrawireless.avphone.model.AvPhoneApplication
import com.sierrawireless.avphone.model.AvPhoneObject
import net.airvantage.model.AirVantageException
import net.airvantage.model.Application
import net.airvantage.model.AvSystem
import net.airvantage.model.User
import net.airvantage.utils.IAirVantageClient
import net.airvantage.utils.Utils
import java.io.IOException
import java.util.*

class ApplicationClient internal constructor(private val client: IAirVantageClient) : IApplicationClient {
    private var currentUser: User? = null
    private var mPhoneName: String? = null

    private val application: Application?
        @Throws(IOException::class, AirVantageException::class)
        get() {
            val applications = client.getApplications(AvPhoneApplication.appType(currentUsername, mPhoneName!!))
            return Utils.first(applications)
        }

    private val currentUsername: String
        get() {
            if (currentUser == null) {
                try {
                    currentUser = client.currentUser
                } catch (e: IOException) {
                    return ""
                } catch (e: AirVantageException) {
                    return ""
                }

            }
            return currentUser!!.email!!
        }

    @Throws(IOException::class, AirVantageException::class)
    override fun ensureApplicationExists(phoneName: String): Application {
        mPhoneName = phoneName
        var application = application

        if (application == null) {
            application = createApplication(phoneName)
            setApplicationCommunication(application.uid!!)
        }
        Log.i(TAG, "ensureApplicationExists: application is " + application.uid)
        return application
    }

    @Throws(IOException::class, AirVantageException::class)
    override fun setApplicationData(applicationUid: String, obj: AvPhoneObject) {
        val data = AvPhoneApplication.createApplicationData(obj)
        client.setApplicationData(applicationUid, data)
    }

    @Throws(IOException::class, AirVantageException::class)
    override fun createApplication(phoneName: String): Application {
        val application = AvPhoneApplication.createApplication(currentUsername, phoneName)
        return client.createApplication(application)
    }

    @Throws(IOException::class, AirVantageException::class)
    override fun setApplicationCommunication(applicationUid: String) {
        val protocols = AvPhoneApplication.createProtocols()
        client.setApplicationCommunication(applicationUid, protocols)
    }

    @Throws(IOException::class, AirVantageException::class)
    override fun addApplication(system: AvSystem, application: Application) {
        client.updateSystem(systemWithApplication(system, application))
    }

    private fun systemWithApplication(system: AvSystem, appToAdd: Application): AvSystem {

        val res = AvSystem()
        res.uid = system.uid
        res.type = system.type
        res.applications = ArrayList()

        var appAlreadyLinked = false

        if (system.applications != null) {
            for (systemApp in system.applications!!) {
                val resApp = Application()
                resApp.uid = systemApp.uid
                res.applications!!.add(resApp)
                if (resApp.uid == appToAdd.uid) {
                    appAlreadyLinked = true
                }
            }
        }
        if (!appAlreadyLinked) {
            val addedApp = Application()
            addedApp.uid = appToAdd.uid
            res.applications!!.add(addedApp)
        }

        return res

    }

    companion object {
        private val TAG = ApplicationClient::class.simpleName
    }
}
